Use `$HOME/.crystal` as the default cache dir.

asterite opened the pull request:
And only keep the 10 most recently used directories. Fixes #2435.

The default directory will now be `$HOME/.crystal` so we don't end up with `.crystal` directories everywhere.

I considered @will's [comment](https://github.com/crystal-lang/crystal/issues/2435#issuecomment-211416861) about ignoring `eval` in this, but since we are keeping 10 most recently used directories, and `eval` is always assigned a same directory, the case where a very used directory is evicted is very unlikely, you must be working on 10 programs at the same time, plus using eval. When you run `crystal spec` a `spec` directory inside the cache is also used. In conclusion, probably the 8 most used programs are kept in this cache, if we consider `eval` and `spec` to always be present, which I think it's still good and it keeps the compiler's logic simple.

kirbyfan64 commented:
:(

I prefer the old behavior, especially since it's a persistent cache. Could there at least be an option to use the old/current behavior, like `--cache-current-dir` or something?

jhass commented on src/compiler/crystal/codegen/cache_dir.cr:
```diff
@@ -0,0 +1,63 @@
+         .reject { |name| name == "." || name == ".." }
+         .map! { |name| File.join(@dir, name) }
+         .select { |dir| Dir.exists?(dir) }
+         .sort_by! { |dir| File.stat(dir).mtime rescue Time.epoch(0) }
```
`stat` is fairly expensive and `exists?` already does one. With usually not more than 11 entries in the target directory it's a micro optimization, but skipping the stats if there's `<= 10` entries in `@dir` might be worth it.

asterite replied:
You are right, but in a non-optimized compiler with more than 10 directories it took about `0.003` seconds or something like that, and since the number will be capped to 10 I think keeping the code clear is more important here (some intermediate arrays are created too, but since this isn't going to be a bottleneck we can use these nice features :-)).

jhass commented on src/compiler/crystal/codegen/cache_dir.cr:
```diff
@@ -0,0 +1,63 @@
+         .sort_by! { |dir| File.stat(dir).mtime rescue Time.epoch(0) }
+         .reverse!
+         .skip(10)
+         .each { |name| `rm -rf "#{name}"` rescue nil }
```
Add a todo for replacing this with `Dir.delete(path, recursive: true)` or alike is implemented?

asterite replied:
I actually searched if we have a method like that before writing this. I couldn't find it so I thought "maybe we are using `rm -rf ...`?" and I was right. When we add `Dir.delete(path, recursive: true)` I'm sure we will remember changing all those `rm -rf ...`. Or it can be added as a reminder for the issue that tracks that feature.

asterite commented:
@kirbyfan64 You would set the `CRYSTAL_CACHE_DIR` environment variable to ".crystal".

I don't think keeping the cache with more than 10 elements makes sense. The cache speeds up things, in some cases it reduces compilation time by half, but if you go back to some old program you didn't touch for a few weeks then you wouldn't mind loosing 5~10 seconds (for a big program like the compiler) just once.

jhass commented:
Let's do a writable check on both `$HOME` and `$HOME/.crystal` when it already exists, stuff like `sudo -s` doesn't always reset it.

jhass commented on src/compiler/crystal/codegen/cache_dir.cr:
```diff
@@ -0,0 +1,63 @@
+      Dir.entries(@dir)
+         .reject { |name| name == "." || name == ".." }
+         .map! { |name| File.join(@dir, name) }
+         .select { |dir| File.file?(dir) }
```
Could extract these four into another method and `partition` into files and directories, `files, directories = gather_cache_files; cleanup_files(files); cleanup_dirs(dirs)`

luislavena commented:
@asterite does this address the `crystal spec` issue? The directory used by `crystal spec` was not bound to the directory used to invoke the command (so is not project-specific).

asterite commented:
@luislavena No, a directory named `spec` is always used for `crystal spec`. I can probably fix that too.

asterite commented:
I made a few changes:
- `CacheDir` is now a singleton because it's used in a couple of places.
- It checks that the cache directory is writable (well, it tries to create it), and if not then it falls back to ".crystal" (relative to the current directory)
- When running `crystal spec`, the directory in the cache is named `current directory + spec` so running `crystal spec` from different projects will use different cache entries
- I did some small refactors like adding `gather_cache_entries`. We could still avoid some small duplicated calls, but I think these would just be micro-optimizations (and I'm also in favor of adding a bit duplication instead of adding a dependency, a Go proverb).

jhass commented on src/compiler/crystal/codegen/cache_dir.cr:
```diff
@@ -0,0 +1,95 @@
+      begin
+        Dir.mkdir_p(@dir)
+      rescue Errno
+        @dir = ".crystal"
```
What if that doesn't exist yet? What if that isn't writable?

asterite replied:
I will `Dir.mkdir_p(".crystal")`, and if that fails too I'll give a compiler error. Sounds good?

jhass replied:
Yes :)

asterite replied:
@jhass Done!

chocolateboy commented:
Shouldn't this use `${XDG_CACHE_HOME:-~/.cache}/crystal`? ([cf.](https://github.com/puniverse/capsule/issues/104))

> `$XDG_CACHE_HOME` defines the base directory relative to which user specific non-essential data files should be stored. If `$XDG_CACHE_HOME` is either not set or empty, a default equal to `$HOME/.cache` should be used.
>
> -- [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.7.html)

ysbaddaden commented:
It would be nice to follow XDG on Linux, yes.

asterite commented:
I updated the PR. Now the [candidate directories](https://github.com/crystal-lang/crystal/pull/2476/files#diff-929557c44b8227b7f89f6ca6e72c7d8eR70) are:

```crystal
candidates = [
  ENV["CRYSTAL_CACHE_DIR"]?,
  ENV["XDG_CACHE_HOME"]?.try { |home| "#{home}/crystal" },
  ENV["HOME"]?.try { |home| "#{home}/.cache/crystal" },
  ENV["HOME"]?.try { |home| "#{home}/.crystal" },
  ".crystal",
]
``` 

The first one that can be created is used.

I also added CRYSTAL_CACHE_DIR to the output of `crystal env`, so you can quickly know where's the cache directory in case you want to inspect/delete it.
